pypeit/utils.py

```diff
@@ -682,50 +713,58 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, inmask = None, upper=5,
-    msgs.info("******************************  Iter  Chi^2  # rejected  Rel. fact   ******************************")
-    msgs.info("                              ----  -----  ----------  --------- ")
+    if ingpm is None:
+        ingpm = invvar > 0
 
+    if not quiet:
+        termwidth = 80-13
+        msgs.info('B-spline fit:')
+        msgs.info('    npoly = {0} profile basis functions'.format(npoly))
+        msgs.info('    ngood = {0}/{1} measurements'.format(np.sum(ingpm), ingpm.size))
+        msgs.info(' {0:>4}  {1:>8}  {2:>7}  {3:>6} '.format(
+                    'Iter', 'Chi^2', 'N Rej', 'R. Fac').center(termwidth, '*'))
+        hlinestr = ' {0}  {1}  {2}  {3} '.format('-'*4, '-'*8, '-'*7, '-'*6)
+        nullval = '  {0:>8}  {1:>7}  {2:>6} '.format('-'*2, '-'*2, '-'*2)
+        msgs.info(hlinestr.center(termwidth))
 
-    maskwork = outmask & inmask & (invvar > 0)
+    maskwork = outmask & ingpm & (invvar > 0)
     if not maskwork.any():
         msgs.error('No valid data points in bspline_profile!.')
-    else:
-        # Init bspline class
-        sset = pydl.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
-                                funcname='Bspline longslit special', **kwargs_bspline)
-        if maskwork.sum() < sset.nord:
+
+    # Init bspline class
+#    sset = pydl.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
+#                            funcname='Bspline longslit special', **kwargs_bspline)
+    sset = bspline.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
+                           funcname='Bspline longslit special', **kwargs_bspline)
+    if maskwork.sum() < sset.nord:
+        if not quiet:
             msgs.warn('Number of good data points fewer than nord.')
-            exit_status = 4
-            return sset, outmask, yfit, reduced_chi, exit_status
+        # TODO: Why isn't maskwork returned?
+        return sset, outmask, yfit, reduced_chi, 4
 
     # This was checked in detail against IDL for identical inputs
-    outer = (np.outer(np.ones(nord, dtype=float), profile_basis.flatten('F'))).T
+    # KBW: Tried a few things and this was about as fast as you can get.
+    outer = np.outer(np.ones(nord, dtype=float), profile_basis.flatten('F')).T
     action_multiple = outer.reshape((nx, npoly * nord), order='F')
     #--------------------
     # Iterate spline fit
     iiter = 0
-    error = -1
-    qdone = False
+    error = -1                  # Indicates that the fit should be done
+    qdone = False               # True if rejection iterations are done
     exit_status = 0
     relative_factor = 1.0
-    tempin = np.copy(inmask)
-    while (error != 0 or qdone is False) and iiter <= maxiter and (exit_status == 0):
+    nrel = 0 if relative is None else len(relative)
+    # TODO: Why do we need both maskwork and tempin?
```

Comment:
It has to do with how djs_reject works. I would need to think throught it carefully and go through that code.

Reply:
Okay.  In infinite time ...